### ISSUE TITLE
possible CW timing improvements

### ISSUE THREAD

**Issue by MikeJohnshoy** [drexjj/sbitx#86]
I am starting to look at CW keying improvements.  
I think the objective should be getting the straight key working up to a "high" speed (30 wpm?).
Electronic keying will get better with straight key improvments.
I don't  think full break-in should be expected - but we need to be responsive enough to catch most of the first dot or dash.  
CW delay should probably be adjusted so that we only switch back to receive after a full character at least (not between dots and dashes). 

NOTE: At 30 wpm a 'dit' is 40 milliseconds - how much of the first dit can we lose?  

Walk with me through the sbitx.c code and think about what's happening!
Imagine the straight key has been pressed - now it's critical that we respond fast!

     void tr_switch_de(int tx_on){

NOTE: Indentation does not clearly show this but the first half of this function turns transmit on, the second half
turns it off

		if (tx_on){
			//mute it all and hang on for a millisecond
			sound_mixer(audio_card, "Master", 0);
			sound_mixer(audio_card, "Capture", 0);
			delay(1);
			
Shutting down the audio is good and necessary.  I assume that delay(1) is 1 msec?
NOTE: this is the first of several delays.  Can they be stacked up with a single overall delay maybe?

			//ADDED BY KF7YDU - Check if ptt is enabled, if so, set ptt pin to high
			if (ext_ptt_enable == 1) {
				digitalWrite(EXT_PTT, HIGH);
			}
			delay(20);
			
NOTE:  we take that 20 msec delay whether ext_ptt_enable is on or not (this is QRO mod I think)
Can we move the delay before the end of that if statement so it won't affect people not using QRO?	

			//now switch of the signal back
			//now ramp up after 5 msecs
			delay(2);
			
NOTE: I don't know what that comment means, or what that delay(2) is for? Does delay(2) mean 5 milliseconds?

			digitalWrite(TX_LINE, HIGH);
			mute_count = 20;
			tx_process_restart = 1;
			//give time for the reed relay to switch
			delay(2);
			
NOTE:  Farhan sent a mod kit to all DE owners that got rid of the T/R relay.  I think we should assume that a DE sbitx has the 
solid state T/R mod in place.  Newer sbitx have also gotten rid of all the relays controlling the bank of low pass filters - DE owners
will still hear those low pass filter relays when changing bands.  Is this delay still needed?

NOTE: Not sure what is happening next, but there is another delay ...

			set_tx_power_levels();
			in_tx = 1;
			//finally ramp up the power

Is tr_relay a variable that tells whether the sbitx user installed the solid state  T/R switch mod?
It is set to 0 at start of sbitx.c, should there be a note up there that user should set this if needed?
Or, if it is always 0 then this is dead code ...

			if (tr_relay){
				set_lpf_40mhz(freq_hdr);
				delay(10); //debounce the lpf relays
			}
			digitalWrite(TX_POWER, HIGH);
			spectrum_reset();
		}

NOTE:  This is the second major half of the tr_switch_de function, where we turn back to receive.
The indentation does not help make this clear
I don't think this is a time critical as the code above

		else {
			in_tx = 0;
			//mute it all and hang on
			sound_mixer(audio_card, "Master", 0);
			sound_mixer(audio_card, "Capture", 0);
			delay(1);
      fft_reset_m_bins();
			mute_count = MUTE_MAX;

			//power down the PA chain to null any gain
			digitalWrite(TX_POWER, LOW);
			delay(2);

NOTE:  What does this next code do?  I see the code for tr_switch_v2 treats the low pass filters differently 
when switch to transmit, and then when switching back to receive

			if (tr_relay){

Is this some variable that tells whether the sbitx user installed the solid state  T/R switch mod?
Could this be unreachable/dead code?

  			digitalWrite(LPF_A, LOW);
  			digitalWrite(LPF_B, LOW);
 	 		digitalWrite(LPF_C, LOW);
  			digitalWrite(LPF_D, LOW);
			}

but apparently we still have to suffer another delay?

			delay(10);

NOTE:  This next code does not include a change needed to restore audio volume that is needed but I don't want to mix
versions ...

			//drive the tx line low, switching the signal path 
			digitalWrite(TX_LINE, LOW);
			digitalWrite(EXT_PTT, LOW); //ADDED by KF7YDU, shuts down ext_ptt. 
			delay(5); 
			//audio codec is back on
            check_r1_volume();
            sound_mixer(audio_card, "Master", rx_vol);
			sound_mixer(audio_card, "Capture", rx_gain);
			spectrum_reset();
I don't know what this last comment is referring to
			//rx_tx_ramp = 10;
    }
}

Bottom line - how many delays can we get out of the switch to transmit?  I think possibly a significant amount

If this walk-through seems productive, we could do the same with tr_switch_v2(). 

There may or may not be a problem with the rpi polling the key, and that is clearly part of the end-to-end timeline. And that all happens before we get into the code here.

I guess there might be a problem with version control doing large scale cleanups of code formatting / indentation
I probably need to just use a smarter editor

WE NEED TO BE CAREFUL WITH THE TRANSMIT POWER CHAIN THAT WE DON'T TAKE OUT SOME CRITICAL SAFETY FEATURE TRYING TO SAVE TIME!

**Comment by drexjj** [comment]
> I guess there might be a problem with version control doing large scale cleanups of code formatting / indentation
I probably need to just use a smarter editor

We are using vs code now with the C plugin (you can find it in C++) to get the format uniformity..

**Comment by drexjj** [comment]
> Bottom line - how many delays can we get out of the switch to transmit? I think possibly a significant amount

If this walk-through seems productive, we could do the same with tr_switch_v2().

This is going to take some time to go through and test. Unfortunately nobody on the development team owns a DE or is proficient in CW ops. Perhaps you can test and provide feedback?

**Comment by MikeJohnshoy** [comment]
Sure I will.  I was trying to decide how one would test this objectively.  Maybe some debug code could physically time the execution from calling tr_switch_de until the transmit switch is complete?  What are we actually starting with - if the first 'dit' coming from the keyer is 40 ms (at 30 wpm) how much are we missing?  Might be cool to have a starting measurement, and then see what each change contributes ...

I should be able to see if it "feels" better.  I am not fast on CW straight key (though I like it) and I don't have a paddle yet (but I might have to get one to test this!)

I was discouraged to see a lot of discussion about CW performance focusing on polling the state of the key or paddle,  increasing the polling rate, or priority, or even making it interrupt driven which seems like it might have big repercussions.  Just making the T/R switch function cleaner/more efficient seems like the way to go, if it gets us far enough.

**Comment by MikeJohnshoy** [comment]
I started with the tr_switch_de and tr_switch_v2 code in upcoming 4.02
I focused on improving timliness for switching to transmit - times returning to receive are dominated
by user setting of delay. I made a few reductions in both transmit and receive, but put the the receive reductions back in to be conservative.  I made minor changes that give a 20+ msec savings on a DE and v2 and newer sbitx going to transmit.  This is a reduction in T/R switching time that should make a modest improvement in CW break-in time.  
I don't think there is much more to gain in reducing T/R switching time (need to look earlier in the keying timeline).

The tr_switch_de code maybe hasn't seen much use for a while.  I did a short test on the latest 4.01 release, and tr_switch_de is never called on my DE - the tr_switch_v2 code gets called instead.  

I would like the two pieces of code to look as similar as possible, but re-organizing things adds
risk - so I didn't re-organize.  In general I think the "de" code should look a lot like
tr_switch_v2 except for controlling the LPFs.  I added a lot of comments as I tried to figure out what was happening.  Let me know if I got things wrong!



[MJ sbitx tr switch next.txt](https://github.com/user-attachments/files/17301407/MJ.sbitx.tr.switch.next.txt)

**Comment by drexjj** [comment]
I will take a look at it this weekend!

**Comment by drexjj** [comment]
Please send me the entire file for comparison and testing.

**Comment by MikeJohnshoy** [comment]
[sbitx tr_switch timing data - Google Sheets.pdf](https://github.com/user-attachments/files/17400498/sbitx.tr_switch.timing.data.-.Google.Sheets.pdf)
[sbitx4.02-DE-Relay_Fix and delay reductions with timing code.txt](https://github.com/user-attachments/files/17400526/sbitx4.02-DE-Relay_Fix.and.delay.reductions.with.timing.code.txt)
Getting rid of possibly unnecessary delays saves some time - I'm not sure how accurate the code I use to time switching time is (I think I should be seeing bigger deltas removing some of these delays).  But percentage-wise it may be meaningful!

**Comment by drexjj** [comment]
Would this be better addressed by creating a software buffer before transmit for CW modes like Nanokeyer or WinKeyer? https://blog.radioartisan.com/arduino-cw-keyer  https://www.n3fjp.com/cwschematic.html  

What you are trying to address is the missing first dit or dah when in keydown, correct?

**Comment by MikeJohnshoy** [comment]
It's an end-to-end problem.  At the start of the timeline, with a straight key, some delay in detecting key down might result in missing the event completely, or turning a dash into a dit, or whatever.  Assuming a paddle is being used - you have more information on whether it was a dit or dah.  If there is a delay in detecting the contact you can have a lead-in or tail adjustment and try to cover it up.    Look at what they do on page 12 ...
https://www.k1elsystems.com/files/WK3_Datasheet_v1.3.pdf
The TR switch happens at the end of the timeline but I thought there might be some easy pickings there.  Take any improvements where you can find them!  

I do wish I could collect some better metrics to quantify improvements...

I'm glad the DE Relay Fix is going in, be interesting to see if anyone notices that (I really don't think there is a downside ...)

**Comment by MikeJohnshoy** [comment]
See the attached files for latest timing data and notes, and modified code that goes in sbitx.c 

I conclude:
- version 4.11 tr_switch cuts off the first 11 msec (25%) of the first "dit" switching to transmit (that's probably acceptable, though bear in mind that that ignores what happens earlier in detecting the first key-down)
- I like the cleaner tr_switch_unified code that eliminates additional delays and switching the LPF during a T/R switch. But from the point of view of maintaining a common baseline for perhaps 1,000 deployed sbitx, it's too hard to decide each change is worth the risk without a more significant performance payoff (and testing on more than one old DE machine!)

[sbitx tr_switch timing data - Google Sheets.pdf](https://github.com/user-attachments/files/17577005/sbitx.tr_switch.timing.data.-.Google.Sheets.pdf)
[tr_switch_unified.txt](https://github.com/user-attachments/files/17577008/tr_switch_unified.txt)

**Comment by drexjj** [comment]
Thanks Mike. Could you do a PR to the cw-fixes branch on here so we can perform a code review. It makes it much easier to go through it for the team.

**Comment by MikeJohnshoy** [comment]
PR created - a first for me.

**Comment by drexjj** [comment]
added your changes to the cw-fixes branch. Thank you!
